### PR TITLE
Simplify reference caching

### PR DESCRIPTION
**Pull request by twrightsman**
Resolves https://github.com/BfArM-MVH/GRZ_QC_Workflow/issues/87

### CODE REVIEW

File: assets/schema_input.json [hunk 1]
```diff
@@ -65,43 +65,43 @@
...
-                "errorMessage": "Aligned reads BAM or CRAM, cannot contain spaces and must have extension '.bam' or '.cram'"
+                "errorMessage": "Path to aligned reads must exist, end in '.bam' or '.cram', and cannot contain spaces."
             },
             "fastp_json": {
                 "type": "string",
                 "format": "file-path",
                 "exists": true,
                 "pattern": "\\S+\\.(json)$",
-                "errorMessage": "fastp JSON report from reads in alignment",
+                "errorMessage": "Path to fastp JSON report must exist, end in '.json', and cannot contain spaces.",
```
Comment (kubranarci) [inline assets/schema_input.json:96]: This now can also accept  FASTPLONG results, would be nice to add to the description.

File: conf/modules.config [hunk 2]
```diff
@@ -31,22 +31,15 @@ process {
             mode: params.publish_dir_mode
         ]
     }
-    withName: SAVE_REFERENCE {
+    withName: 'GRZQC:PREPARE_REFERENCES:(SAMTOOLS_BGZIP|SAMTOOLS_FAIDX|BWAMEM2_INDEX|MINIMAP2_INDEX)' {
+        ext.prefix = { "${meta.id}" }
         publishDir = [
-            path: params.save_reference_path ? params.save_reference_path : params.outdir,
+            path: { "${params.reference_path}/${meta.id}" },
```
Comment (kubranarci) [inline conf/modules.config:37]: meta.id is being reference name here? such as GRCh38 or GRCh37? did you also consider cases where user might use another reference path (then reference name will change)

File: subworkflows/local/prepare_references/main.nf [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    ch_samplesheet_reads_lng
+    ch_genome
+
+    main:
+    // general order of precedence: reference_path -> param -> default
+    def reference_path = params.reference_path
+    def ch_versions = Channel.empty()
+
+    // prepare fasta
+    def defaultFasta = [
```
Comment (kubranarci) [inline subworkflows/local/prepare_references/main.nf:18]: Is this in readme? what we use as base reference?

File: subworkflows/local/prepare_references/main.nf [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    emit:
+    fasta    = ch_fasta
+    fai      = ch_fai
+    bwa      = ch_bwa
+    mmi      = ch_mmi
+    versions = ch_versions
+}
+
+def fasta_cached(id, reference_path) {
```
Comment (kubranarci) [inline subworkflows/local/prepare_references/main.nf:87]: One concern I have about references (I know it was the same before) is losing the reference name (exact file names and the original paths) when we create reference directory. I think we should let user to know about the original file names. What do you think about creating a simple readme.txt with the original file names and the indexing processes the file went through in this pipeline? I am not sure how it would be possible yet it will definitely be useful especially for the non experienced users..and it would be a nice provenance act.

File: workflows/grzqc.nf [hunk 5]
```diff
@@ -39,104 +36,6 @@ workflow GRZQC {
     ch_versions = Channel.empty()
     ch_multiqc_files = Channel.empty()
 
-    //
-    // conflicting parameters
-    //
-    // Error if BWA or FAI is given without FASTA and reference_path is not present
-    if ((params.bwa || params.fai) && !params.fasta && !params.reference_path) {
```
Comment (kubranarci) [inline workflows/grzqc.nf:46]: do we have a check if the references are complete when we use reference_path?

**Review by kubranarci — COMMENTED** [review]
In general I like the new reference structering. Before we approve it, can you also make a sort of stress test with edge cases? like one missing fai file only, or bwa directory only.. I want to make sure if we raise correct errors for the users understanding..

File: subworkflows/local/prepare_references/main.nf (hunk 3 above)
Comment (twrightsman) [reply subworkflows/local/prepare_references/main.nf:18]: Good catch, they are out of date in the README: https://github.com/BfArM-MVH/GRZ_QC_Workflow?tab=readme-ov-file#running-the-pipeline-offline

File: subworkflows/local/prepare_references/main.nf (hunk 4 above)
Comment (kubranarci) [reply subworkflows/local/prepare_references/main.nf:87]: exactly, I am thinking a way of printing that info somewhere maybe as a readme.

File: subworkflows/local/prepare_references/main.nf [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    def ch_fasta_state = ch_genome.branch { id ->
+        cached: reference_path ? fasta_cached(id, reference_path).exists() : false
+        not_cached: true
+    }
+    def ch_fasta_cached = ch_fasta_state.cached.map { id -> [[id: id], fasta_cached(id, reference_path)] }
+    def ch_fasta_not_cached = ch_fasta_state.not_cached.map { id ->
+        [[id: id], file(params.fasta ?: defaultFasta[id])]
+    }
+    SAMTOOLS_BGZIP(ch_fasta_not_cached)
```
Comment (fulaibaowang) [inline subworkflows/local/prepare_references/main.nf:31]: would ch_fasta_state.not_cached / SAMTOOLS_BGZIP step branch always run?

File: workflows/grzqc.nf (hunk 5 above)
Comment (kubranarci) [reply workflows/grzqc.nf:46]: so, this implementation, only checks if reference_path is there or not? and if not it creates everything from input fasta?

File: subworkflows/local/prepare_references/main.nf (hunk 6 above)
Comment (fulaibaowang) [reply subworkflows/local/prepare_references/main.nf:31]: same question for SAMTOOLS_FAIDX and BWAMEM2_INDEX

File: subworkflows/local/prepare_references/main.nf (hunk 3 above)
Comment (kubranarci) [reply subworkflows/local/prepare_references/main.nf:18]: that would be great if you can also update those info

File: subworkflows/local/prepare_references/main.nf [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    fasta    = ch_fasta
+    fai      = ch_fai
+    bwa      = ch_bwa
+    mmi      = ch_mmi
+    versions = ch_versions
+}
+
+def fasta_cached(id, reference_path) {
+    // get path that FASTA is cached to
+    file(reference_path) / id / "${id}.fasta.gz"
```
Comment (fulaibaowang) [inline subworkflows/local/prepare_references/main.nf:89]: is it possible to match fa.gz?

File: conf/modules.config (hunk 2 above)
Comment (kubranarci) [reply conf/modules.config:37]: I think it will work

File: workflows/grzqc.nf (hunk 5 above)
Comment (twrightsman) [reply workflows/grzqc.nf:46]: Currently, it only creates things that don't exist in the reference_path (if a reference_path is given). So if FASTA is deleted but bwamem2 is not then FASTA will be re-bgziped but bwamem2 will not be recreated. I'm open to forcing everything to be regenerated if the FASTA is missing but it would increase complexity. Though it would also prevent many potential bugs

File: docs/usage.md [hunk 8]
```diff
@@ -66,34 +66,40 @@ nextflow run main.nf \
...
+    │   ├── GRCh38.bwt.2bit.64
+    │   └── GRCh38.pac
+    ├── GRCh38.fasta.gz
+    ├── GRCh38.fasta.gz.fai
+    ├── GRCh38.fasta.gz.gzi
+    └── GRCh38.fasta.mmi
 ```
 
-The other option is to set `--fasta`, `--fai`, `--bwa` individually, or prepare config a file like this:
+A final, more flexible but unrecommended, alternative is providing `--fasta`, `--fai`, `--bwa` individually, or preparing a config file like this:
```
Comment (fulaibaowang) [inline docs/usage.md:102]: I think `--fai`, `--bwa` is not possible anymore in the current workflow?

Comment (twrightsman) [reply docs/usage.md:102]: Sorry @fulaibaowang you were correct earlier about removing the params from `nextflow.config` since they aren't used, I will also remove them from the schema. I don't think we should keep `--fai`, `--bwa`, and `--mmi`.

**Review by fulaibaowang — APPROVED** [review]
